### PR TITLE
feat(linter): no-empty

### PR DESCRIPTION
**Pull request by Boshen**

### CODE REVIEW

File: crates/oxc_linter/src/rules.rs [hunk 1]
```diff
@@ -1,19 +1,23 @@
...
 }
 
 impl RuleEnum {
     pub fn run<'a>(&self, kind: AstKind<'a>, ctx: &LintContext<'a>) {
         match self {
             Self::NoDebugger(rule) => rule.run(kind, ctx),
+            Self::NoEmpty(rule) => rule.run(kind, ctx),
         }
     }
 }
```
Comment (Boshen) [inline crates/oxc_linter/src/rules.rs:23]: TODO: macro this

File: crates/oxc_linter/src/rules/no_empty.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+#[derive(Debug, Default, Clone)]
+pub struct NoEmpty;
+
+impl Rule for NoEmpty {
+    fn run<'a>(&self, kind: AstKind<'a>, ctx: &LintContext<'a>) {
+        match kind {
+            AstKind::BlockStatement(block) if block.body.is_empty() => {
+                // TODO: check comment
+                ctx.diagnostic(NoEmptyDiagnostic("block", block.span));
```
Comment (Boshen) [inline crates/oxc_linter/src/rules/no_empty.rs:22]: TODO: check comment

Comment (Boshen) [reply crates/oxc_linter/src/rules/no_empty.rs:22]: TODO: check if the parent block is a catch clause

File: crates/oxc_linter/src/rules/no_empty.rs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+use crate::{context::LintContext, rule::Rule};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint(no-empty): Disallow empty block statements")]
+#[diagnostic(severity(warning), help("Add comments inside {0} statement"))]
+struct NoEmptyDiagnostic(&'static str, #[label("Empty {0} statement")] pub Span);
+
+#[derive(Debug, Default, Clone)]
+pub struct NoEmpty;
```
Comment (Boshen) [inline crates/oxc_linter/src/rules/no_empty.rs:15]: TODO: macro this so we get nice documentation out of this. Checkout how Rome / Ruff / Clippy does it.

Comment (Boshen) [reply crates/oxc_linter/src/rules/no_empty.rs:15]: TODO: allowEmptyCatch

File: crates/oxc_linter/src/rules/no_empty.rs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+                    return;
+                }
+                ctx.diagnostic(NoEmptyDiagnostic("block", block.span));
+            }
+            // The visitor does not visit the `BlockStatement` inside the `CatchClause`.
+            // See `Visit::visit_catch_clause`.
+            AstKind::CatchClause(catch_clause)
+                if !self.allow_empty_catch && catch_clause.body.body.is_empty() =>
+            {
+                if ctx.semantic().trivias().has_comments_between(catch_clause.body.span) {
```
Comment (thepassle) [inline crates/oxc_linter/src/rules/no_empty.rs:47]: Do I understand correctly that if the `body` of a `CatchClause` has comments, and you'd check the `body` to find comments, you won't find them as AST Nodes?

File: crates/oxc_linter/src/rules/no_empty.rs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+pub struct NoEmpty {
+    allow_empty_catch: bool,
+}
+
+const RULE_NAME: &str = "no-empty";
+
+impl Rule for NoEmpty {
+    const NAME: &'static str = RULE_NAME;
+
+    fn from_json(value: serde_json::Value) -> Self {
```
Comment (thepassle) [inline crates/oxc_linter/src/rules/no_empty.rs:24]: its not immediately clear to me what this does exactly?

File: crates/oxc_linter/src/rules/no_empty.rs (hunk 4 above)
Comment (Boshen) [reply crates/oxc_linter/src/rules/no_empty.rs:47]: Yeah the AST slightly different, and the visitor is not visiting the block body.

File: crates/oxc_linter/src/rules/no_empty.rs (hunk 5 above)
Comment (Boshen) [reply crates/oxc_linter/src/rules/no_empty.rs:24]: Good catch, let me think about this for a bit. It's a constructor for initializing from a json configuration.

**Pull request merged**